ddtrace/tracer.py

```diff
@@ -761,7 +761,13 @@ def set_tags(self, tags):
     def shutdown(self, timeout=None):
         """Shutdown the tracer.
 
-        This will stop the background writer/worker and flush any finished traces in the buffer.
+        This will stop the background writer/worker and flush any finished
+        traces in the buffer.
+
+        Note that the tracer *cannot* be used after it has been shutdown. Any
+        traces generated after a shutdown will result in a ``RuntimeError``
+        being raised on ``span.finish()`` (or when the span context manager
+        exits).
```

Comment:
This happens because `span.finish()` calls `write()`. This means that the tracer will build span objects only to discover at the end that it cannot write them to the agent. Would it make sense to add a quick check in `__enter__` to avoid building spans if we already know we are going to fail? E.g.
~~~ python
def __enter__(self):
    if not self.tracer.writer.is_alive():
        raise RuntimeError(...)
~~~
or do we expect a writer to eventually become available by the time the spans need to be written out?

Reply:
Right. Agreed that it's not ideal to fail only when the trace is finished. This might give the appearance that it works (until it doesn't).

> or do we expect a writer to eventually become available by the time the spans need to be written out?

Yeah, in the case of `AgentWriter`, it does not start the thread until the first trace is received. We'd need a separate `is_shutdown` or similar property to do a check like this.

Since the error here is using a tracer after `shutdown()` has been called, maybe it makes sense to address the fault directly with a `is_shutdown` attribute to the tracer which is checked in `start_span()`/`trace()`. WDYT?

Reply:
The `is_shutdown` attribute or similar would work; however, now that I look at the bigger picture, if this happens only once at shutdown perhaps the single `RuntimeError` can be tolerated? My main concern was if somehow one could inadvertently try to create traces after the tracer has been shut down, which would then result in traces being created and then thrown away.